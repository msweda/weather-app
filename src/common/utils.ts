import createBrowserHistory from 'history/createBrowserHistory'

export const browserHistory = createBrowserHistory()

export function isGeolocationAvailable(): boolean {
  return 'geolocation' in navigator
}

export function getCurrentPosition() {
  return new Promise<Position>((resolve, reject) => {
    if (!isGeolocationAvailable()) {
      throw new Error('Geolocation is unavailable')
    }
    navigator.geolocation.getCurrentPosition(
      position => resolve(position),
      error => reject(error)
    )
  })
}
