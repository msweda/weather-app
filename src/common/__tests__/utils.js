import { getCurrentPosition, isGeolocationAvailable } from '../utils'

const mockGeolocation = {
  getCurrentPosition: jest.fn(),
}

describe('common utils', () => {
  describe('isGeolocationAvailable', () => {
    let isGeolocationDefined

    beforeEach(() => {
      isGeolocationDefined = global.navigator.geolocation !== undefined
    })

    it('returns true when navigator.geolocation exists', () => {
      if (!isGeolocationDefined) {
        global.navigator.geolocation = mockGeolocation
      }
      expect(isGeolocationAvailable()).toBeTruthy()
      if (!isGeolocationDefined) {
        delete global.navigator.geolocation
      }
    })

    it('return false when navigator.geolocation does not exist', () => {
      let _geolocation
      if (isGeolocationDefined) {
        _geolocation = global.navigator.geolocation
        delete global.navigator.geolocation
      }
      expect(isGeolocationAvailable()).toBeFalsy()
      if (isGeolocationDefined) {
        global.navigator.geolocation = _geolocation
      }
    })
  })

  describe('getCurrentPosition', () => {
    let _geolocation

    beforeEach(() => {
      _geolocation = global.navigator.geolocation
    })

    afterEach(() => {
      global.navigator.geolocation = _geolocation
    })

    it('rejects when geolocation is not defined', () => {
      if (global.navigator.geolocation !== undefined) {
        delete global.navigator.geolocation
      }
      expect.assertions(1)
      return expect(getCurrentPosition()).rejects.toThrowErrorMatchingSnapshot()
    })

    it('resolves with position when geolocation.getCurrentPosition succeeds', () => {
      const mockPosition = {
        coords: {
          latitude: 51.1,
          longitude: 45.3,
        },
      }
      mockGeolocation.getCurrentPosition.mockImplementationOnce(success =>
        success(mockPosition)
      )
      global.navigator.geolocation = mockGeolocation
      expect.assertions(1)
      return expect(getCurrentPosition()).resolves.toEqual(mockPosition)
    })

    it('rejects when geolocation.getCurrentPosition fails', () => {
      const mockError = new Error()
      mockGeolocation.getCurrentPosition.mockImplementationOnce((_, failure) =>
        failure(mockError)
      )
      global.navigator.geolocation = mockGeolocation
      expect.assertions(1)
      return expect(getCurrentPosition()).rejects.toEqual(mockError)
    })
  })
})
