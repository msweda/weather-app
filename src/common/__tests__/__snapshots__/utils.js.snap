// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common utils getCurrentPosition rejects when geolocation is not defined 1`] = `"Geolocation is unavailable"`;
